nomad/event_sink_manager.go

```diff
@@ -0,0 +1,565 @@
+// to the managedsink's SinkWriter
+type ManagedSink struct {
+	// stopCtx is the passed in ctx used to signal that the ManagedSink should
+	// stop running
+	stopCtx context.Context
+
+	// Sink is the state store EventSink
+	Sink structs.EventSink
+
+	// watchCh is used to watch for updates to the ManagedSink's Sink.
+	watchCh <-chan error
+
+	// doneReset is used to notify that the ManagedSink is done reloading
+	// itself from a subscription or state store change
+	doneReset chan struct{}
+
+	// Subscription is the event stream Subscription
+	Subscription *stream.Subscription
+
+	// lastSuccess is the index of the last successfully sent index
+	lastSuccess uint64
+
+	// SinkWriter is an interface used to send events to their final destination
+	SinkWriter stream.SinkWriter
+
+	// stateFn returns the current server's StateStore
+	stateFn func() *state.StateStore
+
+	// broker is the current server's event broker
+	broker *stream.EventBroker
+
+	// sinkCtx is used to signal that the sink needs to be reloaded
+	sinkCtx context.Context
+
+	// cancelFn cancels sinkCtx
+	cancelFn context.CancelFunc
+
+	// mu coordinates access to running
+	mu sync.Mutex
+
+	// running specifies if the managed sink is running
+	running bool
+
+	l hclog.Logger
+}
+
+// NewManagedSink returns a new ManagedSink for a given sinkID. It queries the
+// state store and subscribes the sink to the state stores event broker
+func NewManagedSink(ctx context.Context, sinkID string, stateFn func() *state.StateStore, L hclog.Logger) (*ManagedSink, error) {
+	state := stateFn()
+	if state == nil {
+		return nil, fmt.Errorf("unable to fetch state store")
+	}
+
+	if L == nil {
+		return nil, fmt.Errorf("logger was nil")
+	}
+
+	ws := state.NewWatchSet()
+	sink, err := state.EventSinkByID(ws, sinkID)
+	if err != nil {
+		return nil, fmt.Errorf("getting sink %s: %w", sinkID, err)
+	}
+
+	// TODO(drew) generate writer based off type
+	writer, err := stream.NewWebhookSink(sink)
+	if err != nil {
+		return nil, fmt.Errorf("generating sink writer for sink %w", err)
+	}
+	broker, err := state.EventBroker()
+	if err != nil {
+		return nil, err
+	}
+
+	sinkCtx, cancel := context.WithCancel(ctx)
+	ms := &ManagedSink{
+		stopCtx:    ctx,
+		Sink:       *sink,
+		watchCh:    ws.WatchCh(sinkCtx),
+		doneReset:  make(chan struct{}),
+		SinkWriter: writer,
+		broker:     broker,
+		cancelFn:   cancel,
+		sinkCtx:    sinkCtx,
+		stateFn:    stateFn,
+		l:          L.Named("managed_sink"),
+	}
+
+	req := &stream.SubscribeRequest{
+		Topics: ms.Sink.Topics,
+		Index:  ms.Sink.LatestIndex,
+	}
+
+	sub, err := ms.broker.Subscribe(req)
+	if err != nil {
+		return nil, fmt.Errorf("unable to subscribe sink %w", err)
+	}
+	ms.Subscription = sub
+
+	return ms, nil
+}
+
+// Run runs until the ManagedSink returns an non reloadable error or until the
+// parent ctx is stopped.
+func (m *ManagedSink) Run() error {
+	m.mu.Lock()
+	if m.running {
```

Comment:
There's a bunch of little state-tracking you're needing to do around `m.running` and I wonder if it might be easier to move the starting of the loop into the `NewManagedSink` and then have a `Run/Reload` that resets the watch.